MF - 1416 - Queue Abstraction for Mainflux & RabbitMQ Support

rodneyosodo opened the pull request:
### What does this do?
Changes the messaging broker from NATS to RabittMQ

### Which issue(s) does this PR fix/relate to?
Resolves #1416 

### List any changes that modify/break current functionality

### Have you included tests for your changes?
Not yet. Will include after approval

### Did you document any new/modified functionality?
Not yet

### Notes
N/A

manuio committed:
MF-1263 - Move repeating errors to the separate package (#1540)

* MF-1263 - Mv duplicated errors to pkg/errors

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Revert test build flags

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix merge

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix comment

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

Co-authored-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

manuio committed:
NOISSUE - Fix auth members list response (#1555)

* NOISSUE - Fix auth members list response

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Move group type next to page details

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Rm membersRes

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix typo

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

manuio committed:
MF-1261 - Use StatusUnauthorized for authn and StatusForbidden for authz (#1538)

* MF-1261 - Use StatusUnauthorized for authn and StatusForbidden for authz

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* ErrExternalKey typo

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Rename ErrUnauthorizedAcces -> ErrAuthentication

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix bootstrap error

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix status code in openapi

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix test description

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix test description

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix test description

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Add errors cases

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix status codes

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Add gRPC stutus code

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix tests description

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix openapi and encodeError

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix grpc message

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix test descriptions

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Revert sdk error

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix typo

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

blokovi committed:
MF-1059 - Add TLS support for email (#1560)

* Use gomail package for sending emails

Signed-off-by: Ivan Milosevic <iva@blokovi.com>

* remove print err

Signed-off-by: Ivan Milosevic <iva@blokovi.com>

* Add vendor

Signed-off-by: Ivan Milosevic <iva@blokovi.com>

* Rename email structure
remove logger

Signed-off-by: Ivan Milosevic <iva@blokovi.com>

* typo in var name

Signed-off-by: Ivan Milosevic <iva@blokovi.com>

* rename var

Signed-off-by: Ivan Milosevic <iva@blokovi.com>

* remove MF_EMAIL_SECRET

Signed-off-by: Ivan Milosevic <iva@blokovi.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Initial commit of adding rabbitmq broker

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Initial commit of adding rabbitmq broker

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Initial commit of adding rabbitmq broker

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

tzzed committed:
NOISSUE - Refactor MQTT subscriber (#1561)

* correct suscriber interface validator + refactore token error handling

Signed-off-by: tzzed <zerouali.t@gmail.com>

* apply review suggestion

Signed-off-by: tzzed <zerouali.t@gmail.com>

Co-authored-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Initial commit for tests

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Initial commit for tests

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

mteodor committed:
MF-1257 - Access messages from readers endpoint with user access token (#1470)

* remove owner id

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

* add user auth for db reader

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* add user auth for db reader

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* enable mongodb reader for user token reading

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* use uuid check for auth switch between thing key and user tok

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* enable user token reading

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* revert to correct version

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* fix endpoint test, add additional tests

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* remove logs,dead code

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* fix logging messages

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* remove auth interface, add authorization header type

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* update api doc

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* remove unused package

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* some refactor of cases for authorization switch

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* correct description in openapi

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* fix endpoint test to match auth service change

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* some rename

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* initialize auth url

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* add env variables for auth service

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* fix spelling

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* Things prefix and no prefix for Thing authorization, Bearer for user

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* update readme file

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* fix default things grpc port

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* enable user reading for timescaledb

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* remove not used error

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* improve errors

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* refactor authorize

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* add chanID check

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* inline some error checking

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* fixing errors

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* fixing errors

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* improve test case description

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* remove test code

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* dont inline

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* refactor a bit encodeError

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* remove unused error

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* remove unused error

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* fix things auth grpc url

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* rename variables for header prefix

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

Co-authored-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Bump up tests

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Bump up tests

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Add more tests

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Add more tests

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Add go routines

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Add go routines

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

manuio committed:
MF-1551 - Fix Cobra usage commands and clean unnecessary struct types (#1558)

* MF-1551 - Fix Cobra usage commands and clean unnecessary struct types

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Use linux syntax for cmd usage description

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix typo

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix cmd.Use

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

blokovi committed:
NOISSUE - Separate Keto hosts for read and write (#1563)

* Separate keto hosts for read and write

Signed-off-by: Ivan Milosevic <iva@blokovi.com>

* update readme with new envars

Signed-off-by: Ivan Milosevic <iva@blokovi.com>

* rename read connection name

Signed-off-by: Ivan Milosevic <iva@blokovi.com>

Co-authored-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>
Co-authored-by: Drasko DRASKOVIC <drasko.draskovic@gmail.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Fix tests

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

manuio commented on pkg/messaging/rabbitmq/pubsub.go:
```diff
@@ -77,7 +77,7 @@ func NewPubSub(url string, logger log.Logger) (PubSub, error) {
 		logger:        logger,
 		subscriptions: make(map[string]bool),
-		done:          make(chan error),
+		done: make(chan bool),
```
Looks like the indentation is worng here

rodneyosodo replied:
Resolved

manuio commented on pkg/messaging/rabbitmq/pubsub.go:
```diff
@@ -176,7 +175,6 @@ func (ps *pubsub) handle(deliveries <-chan amqp.Delivery, done chan error, h mes
-	done <- nil
-}
+	<- done 
+}
```
new line at the end of the file please

rodneyosodo replied:
Resolved

dborovcanin committed:
Update dependencies (#1564)

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Fix with wait groups

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
unsubscribe to stop delivering messages

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

manuio committed:
MF-1240 - Return to service transport layer only service errors (#1559)

* MF-1240 - Return to service transport layer only service errors

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Remove unecessary errors

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Rm duplicated errors and fix transport

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Revert http endpoint_test

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix conflict

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

Co-authored-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Remove exclusivity

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Implement cancel mechanisms

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

bioinformatx committed:
MF-1469 - Indicate proper authentication scheme in Authorization header (#1523)

* MF-1469 - Indicate proper authentication scheme in Authorization header

Signed-off-by: Stefan Kovacevic <jen2tri@gmail.com>

* Fixing the remarks on the last push

Signed-off-by: Stefan Kovacevic <jen2tri@gmail.com>

* Remove Bearer prefix in all services and fix tests

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix remarks

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

Co-authored-by: Manuel Imperiale <manuel.imperiale@gmail.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Queuename as parameter

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Queuename as parameter

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
NOISSUE - Add nats wrapper for COAP (#1569)

* Add nats wrapper for COAP

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* Pass pubsub as argument

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* Defer close connection

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* Defer close connection

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* Rename endpoint to topic

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

manuio committed:
MF-1348 - Add transport errors logging (#1544)

* MF-1348 - Add go-kit transport level logging

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix reviews

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix reviews

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix merge

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix remark

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix go test flags

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Use httputil errors in things and http service

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix SDK tests

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Use httputil errors in certs and provision service

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Use httputil errors in consumers service

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* General renaming and add ErrMissingToken

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Rename httputil -> apiutil and use errors in users servive

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Use apiutil errors in auth, bootstrap, readers, things and twins

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Replace errors.Contain by comparison

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix remarks

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Simplify validateID

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Simplify validateID

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Simplify and rename ExtractAuthToken -> ExtractBearerToken

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix readers

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix auth key test and remarks

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Improve comment

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Simplify validateUUID check

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix typo

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

Co-authored-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

manuio committed:
MF-1567 - Use Bearer, Thing or Basic scheme in Authorization header (#1568)

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

manuio committed:
MF-1565 - Document Bearer, Thing and Basic Authorization header (#1566)

* MF-1565 - Document Bearer Authorization header

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix auth, bootstrap, http and readers openapi

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix openapi

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Add enc key for bootstrap

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix typo

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Use global security

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix bearer formats

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Polish descriptions

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix boostrap and typo

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

Co-authored-by: Drasko DRASKOVIC <drasko.draskovic@gmail.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

ibalboteo committed:
MF-1575 Add 'Name' field to ListMembers response in things svc (#1576)

Signed-off-by: Ivan Balboteo <ivan@submer.com>

Co-authored-by: Ivan Balboteo <ivan@submer.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

Hasan98-git committed:
MF-1580 - Influxdb Writer changes format of update-time to string (#1581)

* - MF-1580
- Modified consumers/writers/influxdb/fields.go
- influxdb-writer used to update data type of update-time to string
- Commented line 12 of consumers/writers/influxdb/fields.go to resolve
  uneccessary data type conversion issue

Signed-off-by: Hasan Tariq <hasantariqashraf@gmail.com>

* - MF-1580
- Removed strconv package from consumers/writers/influxdb/fields.go since it is no longer needed
- Removed line 12 from consumers/writers/influxdb/fields.go
- Replaced retrun value of updateTime with msg.UpdateTime (line 16 in
  fields.go)

Signed-off-by: Hasan Tariq <hasantariqashraf@gmail.com>

* Fix InflxuDB readers

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

Co-authored-by: Hasan Tariq <hasant@plcgroup.com>
Co-authored-by: dusanb94 <dusan.borovcanin@mainflux.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

manuio committed:
NOISSUE - Unify MF_INFLUX_READER_DB_HOST and MF_INFLUX_WRITER_DB_HOST envars (#1585)

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

dborovcanin committed:
NOISSUE - Fix CoAP adapter (#1572)

* Revert "NOISSUE - Add nats wrapper for COAP (#1569)"

This reverts commit cc5d5195ab27fa94270ada616487b7053fd9c7bd.

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Fix CoAP adapter

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Update CoAP observation cancel

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Fix observe

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Fix GET handling

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Revert authorization

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Use constants instead of magic numbers

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Remove an empty line

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Extract special observe value to constant

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

manuio committed:
MF-1582 - Fix lora-adapter MQTT client (#1583)

* MF-1582 - Fix lora-adapter MQTT clien

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Add timeout config to the mqtt subscriber

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Rm comment

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Add sub timeout

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

dborovcanin committed:
NOISSUE - Update changelog and readme for release 0.13.0 (#1592)

* Update release example

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Update changelog and examples for 0.13.0 release

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

dborovcanin committed:
Update VerneMQ release (#1593)

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

dborovcanin committed:
NOISSUE - Update changelog for release 0.13.0 (#1595)

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

dborovcanin commented on pkg/messaging/rabbitmq/doc.go:
```diff
@@ -0,0 +1,11 @@
+// Copyright (c) Mainflux
+// SPDX-License-Identifier: Apache-2.0
+
+// Package rabbitmq hold the implementation of the Publisher and PubSub
```
...holds the implementation....

rodneyosodo replied:
Done

dborovcanin commented on pkg/messaging/rabbitmq/pubsub.go:
```diff
@@ -0,0 +1,182 @@
+	QueueWait          = false
+	ConsumerTag        = "mainflux-consumer"
+	Mandatory          = false
+	Immediate          = false
```
Is there any reason these constants are exported?

rodneyosodo replied:
No reason. I am changing that

rodneyosodo replied:
Done

dborovcanin commented on pkg/messaging/rabbitmq/pubsub.go:
```diff
@@ -0,0 +1,182 @@
+const (
+	ChansPrefix        = "channels"
+	SubjectAllChannels = "channels.>"
+	RoutingKey         = "application specific routing key for fancy topologies"
```
Please change the value of the `routingKey`.

rodneyosodo replied:
Done

dborovcanin commented on pkg/messaging/rabbitmq/pubsub.go:
```diff
@@ -0,0 +1,182 @@
+	go func() {
+		ps.handle(msgs, handler)
+		wg.Done()
+	}()
```
There is no `wg.Wait`. What's the purpose of this WaitGroup?

rodneyosodo replied:
Removed the waitgroup. The wait command was blocking

dborovcanin commented on pkg/messaging/rabbitmq/pubsub.go:
```diff
@@ -0,0 +1,182 @@
+	queue         amqp.Queue
+	channel       *amqp.Channel
+	subscriptions map[string]bool
+	done          chan bool
```
Is this used?

rodneyosodo replied:
Yes

dborovcanin replied:
What for, what's the purpose of the `done` channel?

rodneyosodo replied:
When we finish processing messages from the delivery channel we send a nil value

dborovcanin commented on pkg/messaging/rabbitmq/pubsub.go:
```diff
@@ -0,0 +1,182 @@
+		wg.Done()
+	}()
+
+	ps.subscriptions[topic] = true
```
Should this map be protected for concurrent read/write?

rodneyosodo replied:
How?

dborovcanin replied:
Usually by using a mutex.

rodneyosodo replied:
Done

rodneyosodo committed:
unexport constants

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Change routingkey

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Remove wait groups

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
protecting map

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

dborovcanin commented on pkg/messaging/rabbitmq/pubsub.go:
```diff
@@ -0,0 +1,183 @@
+	return ret, nil
+}
+
+func (ps *pubsub) Publish(topic string, msg messaging.Message) error {
```
There is no need to reimplement this in `pubsub`, you can use the implementation from the `publisher.go`:
```go

type pubsub struct {
	publisher
	logger log.Logger
	queue  amqp.Queue
	subscriptions map[string]bool
	mutex sync.Mutex
}
```

rodneyosodo replied:
Good idea

rodneyosodo replied:
I am getting this error `panic: runtime error: invalid memory address or nil pointer dereferenc`. I have resorted to another approach

dborovcanin commented on pkg/messaging/rabbitmq/publisher.go:
```diff
@@ -0,0 +1,80 @@
+
+type publisher struct {
+	connection *amqp.Connection
+	channel    *amqp.Channel
```
Since this is rather a simple struct, consider using `conn` and `ch` instead.

rodneyosodo replied:
Done

rodneyosodo replied:
Done

dborovcanin commented on pkg/messaging/rabbitmq/pubsub.go:
```diff
@@ -0,0 +1,183 @@
+	if err := ps.channel.QueueBind(ps.queue.Name, routingKey, subject, false, nil); err != nil {
+		return err
+	}
+	msgs, err := ps.channel.Consume(ps.queue.Name, "", true, false, false, false, nil)
```
Is it OK to have an empty string for a consumer here?

rodneyosodo replied:
An empty string will cause the library to generate a unique identity for each consumer. Since we have no way to id the subscriber, I think that is the best fit for now.

dborovcanin commented on pkg/messaging/rabbitmq/pubsub.go:
```diff
@@ -0,0 +1,183 @@
+}
+
+func (ps *pubsub) Unsubscribe(topic string) error {
+	defer ps.channel.Cancel("", false)
```
How does canceling for an empty consumer work? Did you test subscribing with multiple consumers and canceling one of them (add that in test cases, as well)?

rodneyosodo replied:
Initially, I thought it would generate the id for the exact consumer but after going through the code it doesn't as referenced [here](https://github.com/rabbitmq/amqp091-go/blob/84fae140a2b7e958129da81bd7143bd93e289520/consumers.go#L100)

rodneyosodo replied:
If we do a `ps.ch.close()` we close the whole channel so we can't subscribe to it. The best solution to this is if we are able to generate UUID so as to id the consumers so we can cancel the respective consumers. The current implementation works fine.

dborovcanin commented on pkg/messaging/rabbitmq/pubsub_test.go:
```diff
@@ -0,0 +1,243 @@
+	data    = []byte("payload")
+)
+
+func TestPubsub(t *testing.T) {
```
If you take publisher implementation from `publisher.go` that's already tested in `publisher_test.go`, you can have only `Subscriber` (with `Subscribe` and `Unsubscribe`) methods tested here.

rodneyosodo replied:
Implemented

dborovcanin commented on pkg/messaging/rabbitmq/pubsub_test.go:
```diff
@@ -0,0 +1,243 @@
+)
+
+func TestPubsub(t *testing.T) {
+	pubsubcases := []struct {
```
Fix typos: `sub topic` -> `subtopic` and `Unsusbcribe` -> `Unsubscribe`.

rodneyosodo replied:
Done

rodneyosodo committed:
Add publisher to pubsub

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Change proto library

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Fix typos

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Reduce pubsub tests based on implementation

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Remove channel cancel

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Export constant

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

fuzhy committed:
NOISSUE - Move invariant statements out of loop for cassandra-writer (#1596)

Signed-off-by: fuzhy <fuzhy1997@outlook.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Embedding publisher into pubsub

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

dborovcanin committed:
NOISSUE - Fix Nginx entrypoint script (#1597)

* Fix Nginx entrypoint script

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Update dependencies

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Fix NginX entrypoint

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Revert Makefile changes

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

drasko commented on pkg/messaging/nats/pubsub.go:
```diff
@@ -36,7 +36,7 @@ type PubSub interface {
 
 type pubsub struct {
-	conn          *broker.Conn
+	publisher
```
Needs variable name

dborovcanin replied:
@drasko I suggested using the composition for reasons explained [here](https://github.com/mainflux/mainflux/pull/1562#discussion_r868436254).

arvindh123 committed:
MF-1525 - Add graceful stop for HTTP and GRPC servers (#1548)

* Add : errgroup to cmd/auth

Signed-off-by: Arvindh <arvindh91@gmail.com>

* Add : Handle graceful stop for auth service
Remove : errgroups from auth service

Signed-off-by: Arvindh <arvindh91@gmail.com>

* Add : Wait till server shutdown

Signed-off-by: Arvindh <arvindh91@gmail.com>

* Change : instead of waitgroup changed to errgroups

Signed-off-by: Arvindh <arvindh91@gmail.com>

* change : KillSignalHandler return type to error

Signed-off-by: Arvindh <arvindh91@gmail.com>

* Empty Commit

Signed-off-by: Arvindh <arvindh91@gmail.com>

* Add : Context to http server shutdown
Rename : varaible from proto to protocol

Signed-off-by: Arvindh <arvindh91@gmail.com>

* change : to default log level

Signed-off-by: Arvindh <arvindh91@gmail.com>

* Add : Sign-off

Signed-off-by: Arvindh <arvindh91@gmail.com>

* Add: graceful stop of http and grpc server

Signed-off-by: Arvindh <arvindh91@gmail.com>

* Fix: typos and caps

Signed-off-by: Arvindh <arvindh91@gmail.com>

* Add: Signed-off

Signed-off-by: Arvindh <arvindh91@gmail.com>

* Rename: Func KillSignalHandler to SignalHandler
Add: SIGABRT

Signed-off-by: Arvindh <arvindh91@gmail.com>

* Fix: auth service

Signed-off-by: Arvindh <arvindh91@gmail.com>

* Add: timeout for grpc gracefulstop
Fix: typos

Signed-off-by: Arvindh <arvindh91@gmail.com>

* Add: .vscode folder to git ignore

Signed-off-by: Arvindh <arvindh91@gmail.com>

* change: variable name to stopWaitTime

Signed-off-by: Arvindh <arvindh91@gmail.com>

* remove: .vscode folder

Signed-off-by: Arvindh <arvindh91@gmail.com>

* remove: .vscode from .gitignore

Signed-off-by: Arvindh <arvindh91@gmail.com>

* Add : logger to handlers

Signed-off-by: Arvindh <arvindh91@gmail.com>

* Add : New line at end of .gitignore file

Signed-off-by: Arvindh <arvindh91@gmail.com>

* Fix : variable naming
Add : graceful stop for timescale

Signed-off-by: Arvindh <arvindh91@gmail.com>

* Remove : unsued NATS library from import

Signed-off-by: Arvindh <arvindh91@gmail.com>

* Move: "https" and "https" to moved to const var

Signed-off-by: Arvindh <arvindh91@gmail.com>

* Move: "http" and "https" to moved to const var

Signed-off-by: Arvindh <arvindh91@gmail.com>

* update:  branch with master

Signed-off-by: Arvindh <arvindh91@gmail.com>

Co-authored-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>
Co-authored-by: Drasko DRASKOVIC <drasko.draskovic@gmail.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Naming publisher

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
MF-1588 - Update Subscriber interface (#1598)

* Initial commit

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* Update subscriber interface

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Add tests

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* Add tests

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* check subscription map

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* Check topic id after topic

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* reword description

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* Setup empty queue

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* Change mqtt implementation

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* Switch statements

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* Simplify

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* Change mqtt subscriber

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* Protect subscription map

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* Fix subscription

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* Set client id

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* Format

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* Change delete method

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

Co-authored-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Update rabbitmq subscriber interface

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
MF - 1590 - Fix fetching list of users with a zero limit (#1594)

* Add max and min limit size

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* Format

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* Format

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

Co-authored-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

dborovcanin commented on pkg/messaging/nats/pubsub.go:
```diff
@@ -42,7 +42,7 @@ type subscription struct {
 
 type pubsub struct {
-	conn          *broker.Conn
+	publisher     publisher
```
Instead of having a `publisher` field, consider using a composition of `publisher` like:
```go
type pubsub struct {
	publisher
	logger        log.Logger
	mu            sync.Mutex
	queue         string
	subscriptions map[string]map[string]subscription
}
```
This will enable using `ps.conn` rather than `ps.publisher.conn` later on.

rodneyosodo replied:
That is what we had initially but @drasko had suggested with name it.

I will stick with the earlier changes we had.

dborovcanin commented on pkg/messaging/rabbitmq/publisher.go:
```diff
@@ -0,0 +1,83 @@
+		immediate,
+		amqp.Publishing{
+			Headers:     amqp.Table{},
+			ContentType: "text/plain",
```
Is this type correct? Since our message is binary, it looks like `application/octet-stream` is more suitable for us.

rodneyosodo replied:
Yes `application/octet-stream` is more suitable

dborovcanin commented on pkg/messaging/rabbitmq/pubsub.go:
```diff
@@ -0,0 +1,192 @@
+	SubjectAllChannels = "channels.>"
+	routingKey         = "mainfluxkey"
+	exchange           = "mainflux"
+	exchangeKind       = "fanout"
```
Did you consider using `topic` kind? In that case, we have a queue per channel and the subtopic could be the routing key. This will have an impact on both publisher and subscriber implementation.

rodneyosodo replied:
Yes. That will be a good approach but it poses come challenges because when we do a subscribe the topic preprocessing is done by the client compared to how it is implemented on the publish method.

dborovcanin replied:
I understand that we'll have to disassemble the topic we created before calling Publish/Subscribe methods, but it's worth trying it because topic manipulation is rather simple. Of course, queue manipulation is also something to take care of (creating a queue if it doesn't exist and removing it if there are no messages and subscribers left). I'm only afraid this topic processing is going to be too cryptic - using the first part of the topic as a queue (name), and the rest as a subtopic (routingKey).

rodneyosodo replied:
From [this](https://www.cloudamqp.com/blog/part4-rabbitmq-13-common-errors.html) reference we need to have multiple queues. The current implementation I have pushed is that we have direct exchange rather than topic exchange as direct is faster than topic exchange. We will have a queue per subscription.
Another way to implement it is to have queues based on services such as a coap queue and an MQTT queue. But this makes us not have many queues.

rodneyosodo replied:
## Overview
RabbitMQ is somehow different to Nats and Kafka in that when a publisher is publishing a message on a topic they will publish the message to an exchange. The message will be published with a routing key. This routing key will direct the message to the different queues onto which the subscriber will consume the message. This is how the message flows
```
publisher -> exchange -> routingkey -> queue -> consumer -> subscriber


The publisher publishes the message to the broker.
The exchange receives the message being published and uses the routing key to route the message to a specific queue
The routing key is used to route the message to a specific queue
The queue temporarily stores messages for the consumer before consuming them
```

![image](https://user-images.githubusercontent.com/28790446/168829991-95a7977f-31ee-4289-952f-1ac121abc670.png)


## Exchange
When the exchange is declared we declare it with a specific kind. This means an exchange can either route messages based on topic, directly to queues or using message headers among other methods. We will be interested in the three routing techniques. The number of bindings increases the performance of topic routing. Fanout exchanges are very fast because there is no routing to process, but this changes if they are bound to a large number of queues. Direct exchanges are faster than topic exchanges if you don't need the wild card. We picked direct method as it strikes a balance between performance and reliability.

## Possible implementations

### One exchange one queue
- If you only have one queue, you'll end up repeating the logic of deciding which code should process the message, when RabbitMQ can do it for you with routing.
- RabbitMQ prefers many queues over a single large one
- This option is ruled out

### One exchange multiple queues
- This is the most viable solution currently
- This is because we will have single consumers on one queue
- The advantage of a single topic exchange is that you get the benefits of both a Direct Exchange and a Fanout Exchange depending on the binding key used.
- Further, configuration changes are easy to achieve and can often be done without disrupting any system processing at all

### Multiple exchanges one queue
- This option is ruled out based on having a single queue

### Multiple exchanges multiple queues
- This is the most preferred way of doing it as the system is highly decoupled. 
- If we cross the boundary and have many exchanges and many queues we are at the risk of the processes taking up more memory which may be significant but this really depends
- If the `msg.channel` is not provided the message won't be consumed as we can subscribe to an empty topic

## Wayforward
- Currently, we have implemented Direct as the exchange type. So if a consumer needs to subscribe to multiple topics we would change this implementation to the exchange having topic as the exchange type
- We have not explored exchange-to-exchange architectures

For more reference check out [this article](https://spring.io/blog/2011/04/01/routing-topologies-for-performance-and-scalability-with-rabbitmq/)

rodneyosodo committed:
using publisher composition

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Change contenttype

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
rename topic for publish and subscribe

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

manuio committed:
NOISSUE - Retrieve client key on cert issuing (#1607)

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

zhang-chuanfeng committed:
fix bug (#1604)

Signed-off-by: zhangchuanfeng <654300242@qq.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Change errors to lower case

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Change errors to lower case

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

manuio commented on pkg/messaging/nats/pubsub_test.go:
```diff
@@ -189,7 +189,7 @@ func TestPubsub(t *testing.T) {
 			topic:        fmt.Sprintf("%s.%s", chansPrefix, topic),
 			topicID:      "",
-			errorMessage: errors.New("empty ID"),
+			errorMessage: errors.New("empty id"),
```
I'd create errors in  a package for nats and rabbitmq and use them in the code and tests

rodneyosodo committed:
export errors

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
queue per subscription

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
queue per subscription

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Change routing method

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Direct method with one exchange to many queues, one consumer per queue

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

dborovcanin commented on pkg/messaging/rabbitmq/publisher.go:
```diff
@@ -0,0 +1,84 @@
+			Headers:     amqp.Table{},
+			ContentType: "application/octet-stream",
+			AppId:       "mainflux-publisher",
+			Body:        []byte(data),
```
No need to cast, data is already `[]byte`.

dborovcanin commented on pkg/messaging/rabbitmq/pubsub_test.go:
```diff
@@ -0,0 +1,179 @@
+			pubsub:       false,
+		},
+		{
+			desc:         "Susbcribe to a topic with a subtopic with an ID",
```
Fix typo `Susbcribe`. Check for other typos as well.

dborovcanin commented on pkg/messaging/rabbitmq/pubsub.go:
```diff
@@ -0,0 +1,167 @@
+}
+type pubsub struct {
+	publisher
+	queue         string
```
What's the purpose of this and is it even used?

rodneyosodo replied:
It is not being used

dborovcanin commented on pkg/messaging/rabbitmq/pubsub.go:
```diff
@@ -0,0 +1,167 @@
+	if err != nil {
+		return err
+	}
+	if err := ps.ch.QueueBind(queue.Name, queue.Name, exchangeName, false, nil); err != nil {
```
Is it safer to use topic instead of `queue.Name`? Is it possible that the broker will change the name passed in QueueDeclare?

rodneyosodo replied:
The broker doesn't change the passed name

rodneyosodo committed:
:recycle: Not casting data

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
:pencil2: Fix typo

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
:recycle: remove passed queue name

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

dborovcanin commented on pkg/messaging/rabbitmq/pubsub.go:
```diff
@@ -0,0 +1,166 @@
+	// SubjectAllChannels represents subject to subscribe for all the channels.
+	SubjectAllChannels = "channels.>"
+	exchangeName       = "mainflux-exchange"
+	exchangeKind       = "direct"
```
Is this used?

rodneyosodo replied:
exchangeKind no. The others yes.

dborovcanin commented on pkg/messaging/rabbitmq/pubsub.go:
```diff
@@ -0,0 +1,166 @@
+}
+
+func (ps *pubsub) Unsubscribe(id, topic string) error {
+	defer ps.ch.Cancel(id, false)
```
Should we really cancel unconditionally?

rodneyosodo replied:
From the docs, this is what it says about the Cancel method

> Only use this method to cleanly stop receiving deliveries from the server and cleanly shut down the consumer chan identified by this tag.  Using this method and waiting for the remaining messages to flush from the consumer chan will ensure all messages received on the network will be delivered to the receiver of your consumer chan.

I believe this is what we want to achieve

dborovcanin replied:
We are, but we should first handle the internal logic, then try to cancel and handle the potential errors. What happens if we have the same consumer subscribed to multiple topics? Does it cancel them all (because it's the same consumer ID)?
For example, a nice solution would be to put this in the `cancel` method of `subscriber` when you subscribe, so that you don't even call it explicitly when unsubscribing. Something like:
```go
	s[id] = subscription{
		cancel: func() error {
			if err := ps.ch.Cancel(id, false); err != nil {
				return err
			}
			return handler.Cancel()
		},
	}
```

rodneyosodo committed:
:fire: removing echange kind

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Combine tests

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Refactor unsubscribe method

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Merge branch 'master' of https://github.com/0x6f736f646f/mainflux into MF-1416

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Fix merge conflict

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
:white_check_mark: sub and unsub to dummy topic

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
generate client id from topic and ID

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Rename topicID to clientID

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
update tests

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

dborovcanin commented on pkg/messaging/rabbitmq/pubsub.go:
```diff
@@ -101,14 +101,16 @@ func (ps *pubsub) Subscribe(id, topic string, handler messaging.MessageHandler)
 	s[id] = subscription{
 		cancel: func() error {
-			if err := ps.ch.Cancel(id, false); err != nil {
+			clientID := fmt.Sprintf("%s-%s", topic, id)
```
Reuse `clientID` from L104.

dborovcanin commented on pkg/messaging/nats/pubsub_test.go:
```diff
@@ -27,9 +27,9 @@ var (
 
 func TestPubsub(t *testing.T) {
-	err := pubsub.Subscribe(topicID, fmt.Sprintf("%s.%s", chansPrefix, topic), handler{})
+	err := pubsub.Subscribe(clientID, fmt.Sprintf("%s.%s", chansPrefix, topic), handler{})
```
You can still keep `TestPublish` and `TestSubscribe` with separate cases (but in this single file, that's OK).

rodneyosodo replied:
Here we are subscribing so that we can receive the published messages

dborovcanin commented on pkg/messaging/nats/pubsub_test.go:
```diff
@@ -83,134 +83,134 @@ func TestPubsub(t *testing.T) {
 			pubsub:       true,
 		},
 		{
 			desc:         "Unsubscribe to a topic with an ID",
```
typo `Unsubscribe **from**...`. Same for all unsubscribe cases.

rodneyosodo committed:
Reuse clientID

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Fix typos

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Seperate testpublish and testpubsub

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

dborovcanin committed:
Merge branch 'master' into MF-1416

drasko reviewed (approved):
LGTM
